build/patches/Enable-crash-reporting-ui.patch

```diff
@@ -27,13 +58,101 @@ diff --git a/chrome/browser/crash_upload_list/crash_upload_list_android.cc b/chr
++
+   std::vector<base::FilePath> allowlist;
+   std::vector<base::FilePath> all_download_dirs =
+       base::android::GetAllPrivateDownloadsDirectories();
+diff --git a/chrome/browser/ui/BUILD.gn b/chrome/browser/ui/BUILD.gn
+--- a/chrome/browser/ui/BUILD.gn
++++ b/chrome/browser/ui/BUILD.gn
+@@ -571,6 +571,7 @@ static_library("ui") {
+     "//third_party/re2",
+     "//third_party/webrtc_overrides:webrtc_component",
+     "//third_party/zlib",
++    "//third_party/zlib/google:zip",
+     "//ui/accessibility",
+     "//ui/base",
+     "//ui/base:data_exchange",
 diff --git a/chrome/browser/ui/webui/crashes_ui.cc b/chrome/browser/ui/webui/crashes_ui.cc
 --- a/chrome/browser/ui/webui/crashes_ui.cc
 +++ b/chrome/browser/ui/webui/crashes_ui.cc
-@@ -173,8 +173,7 @@ void CrashesDOMHandler::OnUploadListAvailable() {
+@@ -38,6 +38,15 @@
+ #include "google_apis/gaia/gaia_auth_util.h"
+ #include "ui/base/resource/resource_bundle.h"
+ 
++#include "base/logging.h"
++#include "base/files/file_util.h"
++#include "base/files/scoped_temp_dir.h"
++#include "base/task/task_traits.h"
++#include "base/task/thread_pool.h"
++#include "base/android/path_utils.h"
++#include "net/base/filename_util.h"
++#include "third_party/zlib/google/zip.h"
++
+ #if BUILDFLAG(IS_CHROMEOS_ASH)
+ #include "chromeos/dbus/dbus_thread_manager.h"
+ #include "chromeos/dbus/debug_daemon/debug_daemon_client.h"
+@@ -71,6 +80,10 @@ content::WebUIDataSource* CreateCrashesUIHTMLSource() {
+   return source;
+ }
+ 
++constexpr base::TaskTraits kLoadingTaskTraits = {
++    base::MayBlock(), base::TaskPriority::USER_BLOCKING,
++    base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN};
++
+ ////////////////////////////////////////////////////////////////////////////////
+ //
+ // CrashesDOMHandler
+@@ -80,7 +93,7 @@ content::WebUIDataSource* CreateCrashesUIHTMLSource() {
+ // The handler for Javascript messages for the chrome://crashes/ page.
+ class CrashesDOMHandler : public WebUIMessageHandler {
+  public:
+-  CrashesDOMHandler();
++  CrashesDOMHandler(content::WebContents* web_contents);
+   ~CrashesDOMHandler() override;
+ 
+   // WebUIMessageHandler implementation.
+@@ -103,15 +116,20 @@ class CrashesDOMHandler : public WebUIMessageHandler {
+   // Asynchronously requests a user triggered upload. Called from JS.
+   void HandleRequestSingleCrashUpload(const base::ListValue* args);
+ 
++  std::string RequestSingleUpload(const std::string& local_id) const;
++  void RequestSingleUploadCallback(const std::string& local_id, const std::string& filename);
++
+   scoped_refptr<UploadList> upload_list_;
+   bool list_available_;
+   bool first_load_;
++  content::WebContents* web_contents_;
+ 
+   DISALLOW_COPY_AND_ASSIGN(CrashesDOMHandler);
+ };
+ 
+-CrashesDOMHandler::CrashesDOMHandler()
+-    : list_available_(false), first_load_(true) {
++CrashesDOMHandler::CrashesDOMHandler(content::WebContents* web_contents)
++    : list_available_(false), first_load_(true),
++      web_contents_(web_contents) {
```

Comment:
I am concerned about the `web_contents`: what if it has been destroyed by the crash already?

Reply:
(Perhaps that is why they were doing this in Java?)

Reply:
> I don't know why that code is in java rather than native...

ah, I understand.
because it uses Android services, so it can send the log even without chromium active